bytecomp/translcore.ml

```diff
@@ -1283,57 +1295,115 @@ and transl_record loc env fields repres opt_init_expr =
+               | Record_regular _ ->
+                 (* TODO: The following code unnecessarily creates a new
+                    record. Flambda optimises it out, but in general
+                    we can avoid it by merging this creation code with the
+                    code below (that destructs the created record) *)
+                 let pos = Typeopt.adjusted_offset desc in
+                 if not desc.lbl_unboxed then access_init (Pfield pos)
+                 else Typeopt.project_fields_into_a_record
+                        ~src_offset:pos ~src:(Lvar init_id) desc.lbl_size ~loc
+               | Record_unboxed _ -> assert false
+               | Record_float -> access_init (Pfloatfield i)
+             in
+             lam, field_kind
            | Overridden (_lid, expr) ->
                let field_kind = value_kind expr.exp_env expr.exp_type in
                transl_exp expr, field_kind)
         fields
     in
     let ll, shape = List.split (Array.to_list lv) in
+    let descs = List.map fst (Array.to_list fields) in
     let mut =
       if Array.exists (fun (lbl, _) -> lbl.lbl_mut = Mutable) fields
       then Mutable
       else Immutable in
     let lam =
       try
         if mut = Mutable then raise Not_constant;
-        let cl = List.map extract_constant ll in
+        let cl = List.fold_right2 (fun l desc cl ->
+          let c = extract_constant l in
+          match c with
+          | Const_block (0, cl') when desc.lbl_unboxed -> cl' @ cl
+          | _ -> c :: cl
+        ) ll descs []
+        in
         match repres with
-        | Record_regular -> Lconst(Const_block(0, cl))
-        | Record_inlined tag -> Lconst(Const_block(tag, cl))
+        | Record_regular { tag; inline; _; } ->
+          begin match inline with
+          | No_inline | Inlined -> Lconst(Const_block(tag, cl))
+          | Extension -> raise Not_constant
+          end
         | Record_unboxed _ -> Lconst(match cl with [v] -> v | _ -> assert false)
         | Record_float ->
             Lconst(Const_float_array(List.map extract_float cl))
-        | Record_extension ->
-            raise Not_constant
       with Not_constant ->
         match repres with
-          Record_regular ->
-            Lprim(Pmakeblock(0, mut, Some shape), ll, loc)
-        | Record_inlined tag ->
+        | Record_regular { tag; inline; has_unboxed_fields; _; } ->
+          let add_extension_slot_if_required ll shape =
+            match inline with
+            | Extension ->
+              let slot =
+                let path =
+                  let (label, _) = fields.(0) in
+                  match label.lbl_res.desc with
+                  | Tconstr(p, _, _) -> p
+                  | _ -> assert false
+                in
+                transl_path env path
+              in
+              slot :: ll, Pgenval :: shape
+            | No_inline | Inlined -> ll, shape
+          in
+          if not has_unboxed_fields then
```

Comment:
This conditional seems to be the reason why the `has_unboxed_fields` property is kept in the record_representation.  I think I would prefer to have the check be done here explicitly (it would be very cheap) instead of being stored in the Types representation (adding redundancy to it).